// hexgate/dist/esm/index.js
/// --- Polyfills ---------------------------------------------------------------
/// Fetch (node-fetch)
import './utils/pollyfills/fetch.js';
export * from './utils/pollyfills/fetch.js';
/// --- Public Modules ----------------------------------------------------------
/// Authentication
export { createHeaders } from './modules/auth/headers.js';
export { auth } from './modules/auth/index.js';
/// Hexgate
export { Hexgate } from './modules/hexgate/index.js';
export { createHexgate } from './modules/hexgate/create.js';
export { extractHexgateSymbol, hexgateSymbol } from './modules/hexgate/symbol.js';
/// Recipes
export { createRecipe, isRecipeFn, CreateWithRecipe } from './modules/recipe/index.js';
export { recipeSymbol, extractRecipeSymbol } from './modules/recipe/symbol.js';
export { RecipeApi } from './modules/recipe/api.js';
/// HTTPS
export { HttpsAgent } from './modules/https/agent.js';
export { createRequestInit } from './modules/https/request-init.js';
/// WebSocket
export { LcuClient, createLcuClient } from './modules/websocket/index.js';
/// Connection
export { Connection } from './modules/connection/index.js';
/// LcuValue
export { LcuValue } from './modules/connection/lcu-value.js';
/// Utilities
export { extractData } from './utils/extract-data.js';
export { identity } from './utils/identity.js';
export { extractDefined } from './utils/extract-defined.js';
export { mapDefined } from './utils/map-defined.js';
export { poll } from './utils/poll.js';
export * from '@cuppachino/proxy-fn';
export { proxyFlyweight } from './utils/proxy-flyweight.js';
export { result } from './utils/result.js';
export { zip } from './utils/zip.js';
export { Observable } from './utils/observable.js';
export { Heartbeat } from './utils/heartbeat.js';
/// Errors
export { HexgateError } from './errors/base-error.js';
export { InvalidPlatformError } from './errors/invalid-platform.js';
export { MissingClientError } from './errors/missing-client.js';
export { PollTimeoutError } from './errors/poll-timeout.js';
