// hexgate/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PollTimeoutError = exports.MissingClientError = exports.InvalidPlatformError = exports.HexgateError = exports.Heartbeat = exports.Observable = exports.zip = exports.result = exports.proxyFlyweight = exports.poll = exports.mapDefined = exports.extractDefined = exports.identity = exports.extractData = exports.LcuValue = exports.Connection = exports.createLcuClient = exports.LcuClient = exports.createRequestInit = exports.HttpsAgent = exports.RecipeApi = exports.extractRecipeSymbol = exports.recipeSymbol = exports.CreateWithRecipe = exports.isRecipeFn = exports.createRecipe = exports.hexgateSymbol = exports.extractHexgateSymbol = exports.createHexgate = exports.Hexgate = exports.auth = exports.createHeaders = void 0;
/// --- Polyfills ---------------------------------------------------------------
/// Fetch (node-fetch)
require("./utils/pollyfills/fetch.js");
__exportStar(require("./utils/pollyfills/fetch.js"), exports);
/// --- Public Modules ----------------------------------------------------------
/// Authentication
var headers_js_1 = require("./modules/auth/headers.js");
Object.defineProperty(exports, "createHeaders", { enumerable: true, get: function () { return headers_js_1.createHeaders; } });
var index_js_1 = require("./modules/auth/index.js");
Object.defineProperty(exports, "auth", { enumerable: true, get: function () { return index_js_1.auth; } });
/// Hexgate
var index_js_2 = require("./modules/hexgate/index.js");
Object.defineProperty(exports, "Hexgate", { enumerable: true, get: function () { return index_js_2.Hexgate; } });
var create_js_1 = require("./modules/hexgate/create.js");
Object.defineProperty(exports, "createHexgate", { enumerable: true, get: function () { return create_js_1.createHexgate; } });
var symbol_js_1 = require("./modules/hexgate/symbol.js");
Object.defineProperty(exports, "extractHexgateSymbol", { enumerable: true, get: function () { return symbol_js_1.extractHexgateSymbol; } });
Object.defineProperty(exports, "hexgateSymbol", { enumerable: true, get: function () { return symbol_js_1.hexgateSymbol; } });
/// Recipes
var index_js_3 = require("./modules/recipe/index.js");
Object.defineProperty(exports, "createRecipe", { enumerable: true, get: function () { return index_js_3.createRecipe; } });
Object.defineProperty(exports, "isRecipeFn", { enumerable: true, get: function () { return index_js_3.isRecipeFn; } });
Object.defineProperty(exports, "CreateWithRecipe", { enumerable: true, get: function () { return index_js_3.CreateWithRecipe; } });
var symbol_js_2 = require("./modules/recipe/symbol.js");
Object.defineProperty(exports, "recipeSymbol", { enumerable: true, get: function () { return symbol_js_2.recipeSymbol; } });
Object.defineProperty(exports, "extractRecipeSymbol", { enumerable: true, get: function () { return symbol_js_2.extractRecipeSymbol; } });
var api_js_1 = require("./modules/recipe/api.js");
Object.defineProperty(exports, "RecipeApi", { enumerable: true, get: function () { return api_js_1.RecipeApi; } });
/// HTTPS
var agent_js_1 = require("./modules/https/agent.js");
Object.defineProperty(exports, "HttpsAgent", { enumerable: true, get: function () { return agent_js_1.HttpsAgent; } });
var request_init_js_1 = require("./modules/https/request-init.js");
Object.defineProperty(exports, "createRequestInit", { enumerable: true, get: function () { return request_init_js_1.createRequestInit; } });
/// WebSocket
var index_js_4 = require("./modules/websocket/index.js");
Object.defineProperty(exports, "LcuClient", { enumerable: true, get: function () { return index_js_4.LcuClient; } });
Object.defineProperty(exports, "createLcuClient", { enumerable: true, get: function () { return index_js_4.createLcuClient; } });
/// Connection
var index_js_5 = require("./modules/connection/index.js");
Object.defineProperty(exports, "Connection", { enumerable: true, get: function () { return index_js_5.Connection; } });
/// LcuValue
var lcu_value_js_1 = require("./modules/connection/lcu-value.js");
Object.defineProperty(exports, "LcuValue", { enumerable: true, get: function () { return lcu_value_js_1.LcuValue; } });
/// Utilities
var extract_data_js_1 = require("./utils/extract-data.js");
Object.defineProperty(exports, "extractData", { enumerable: true, get: function () { return extract_data_js_1.extractData; } });
var identity_js_1 = require("./utils/identity.js");
Object.defineProperty(exports, "identity", { enumerable: true, get: function () { return identity_js_1.identity; } });
var extract_defined_js_1 = require("./utils/extract-defined.js");
Object.defineProperty(exports, "extractDefined", { enumerable: true, get: function () { return extract_defined_js_1.extractDefined; } });
var map_defined_js_1 = require("./utils/map-defined.js");
Object.defineProperty(exports, "mapDefined", { enumerable: true, get: function () { return map_defined_js_1.mapDefined; } });
var poll_js_1 = require("./utils/poll.js");
Object.defineProperty(exports, "poll", { enumerable: true, get: function () { return poll_js_1.poll; } });
__exportStar(require("@cuppachino/proxy-fn"), exports);
var proxy_flyweight_js_1 = require("./utils/proxy-flyweight.js");
Object.defineProperty(exports, "proxyFlyweight", { enumerable: true, get: function () { return proxy_flyweight_js_1.proxyFlyweight; } });
var result_js_1 = require("./utils/result.js");
Object.defineProperty(exports, "result", { enumerable: true, get: function () { return result_js_1.result; } });
var zip_js_1 = require("./utils/zip.js");
Object.defineProperty(exports, "zip", { enumerable: true, get: function () { return zip_js_1.zip; } });
var observable_js_1 = require("./utils/observable.js");
Object.defineProperty(exports, "Observable", { enumerable: true, get: function () { return observable_js_1.Observable; } });
var heartbeat_js_1 = require("./utils/heartbeat.js");
Object.defineProperty(exports, "Heartbeat", { enumerable: true, get: function () { return heartbeat_js_1.Heartbeat; } });
/// Errors
var base_error_js_1 = require("./errors/base-error.js");
Object.defineProperty(exports, "HexgateError", { enumerable: true, get: function () { return base_error_js_1.HexgateError; } });
var invalid_platform_js_1 = require("./errors/invalid-platform.js");
Object.defineProperty(exports, "InvalidPlatformError", { enumerable: true, get: function () { return invalid_platform_js_1.InvalidPlatformError; } });
var missing_client_js_1 = require("./errors/missing-client.js");
Object.defineProperty(exports, "MissingClientError", { enumerable: true, get: function () { return missing_client_js_1.MissingClientError; } });
var poll_timeout_js_1 = require("./errors/poll-timeout.js");
Object.defineProperty(exports, "PollTimeoutError", { enumerable: true, get: function () { return poll_timeout_js_1.PollTimeoutError; } });
