// hexgate/dist/esm/modules/recipe/api.js
import { proxyFlyweight as once } from '../../utils/proxy-flyweight.js';
import { proxyFn as wrap } from '@cuppachino/proxy-fn';
import { result } from '../../utils/result.js';
export class RecipeApi {
    /**
     * The build method of a Hexgate instance.
     */
    build;
    constructor(hexgate) {
        this.build = hexgate['build'];
    }
    /**
     * #### Curried
     * Create a new proxy that will call `from` with the supplied arguments, and then call `to` with the result before returning the result.
     *
     * @see [wrap](../../utils/proxy-function.ts)
     */
    wrap = wrap;
    /**
     * Create a new function that extracts the data from the response of the supplied function.
     */
    unwrap(fn) {
        return async (...args) => {
            const response = await fn(...args);
            return response.data;
        };
    }
    /**
     * @see [once](../../utils/proxy-flyweight.ts)
     */
    once = once;
    /**
     * @see [result](../../utils/result.ts)
     */
    result = result;
    /**
     * Returns empty arguments.
     * @returns Default arguments for a built request function that takes no arguments.
     */
    from() {
        return [];
    }
    /**
     * Extract the data from the response of the supplied function.
     */
    async to(response) {
        const res = await response;
        return res.data;
    }
}
